storage/pinlogs_blockwise.h

```diff
@@ -0,0 +1,86 @@
+
+#if FLASH_BLOCK_WORDS <= 1
+#error "FLASH_BLOCK_WORDS must be at least 2 to fit the counter and header"
+#endif
+
+#define PIN_LOG_HALFWORDS (((FLASH_BLOCK_WORDS - 1) * sizeof(uint32_t)) / 2)
+
+static uint16_t expand_counter(uint16_t c) {
+  c = ((c << 4) | c) & 0x0f0f;
+  c = ((c << 2) | c) & 0x3333;
+  c = ((c << 1) | c) & 0x5555;
+  c = ((c << 1) | c) ^ 0xaaaa;
+  return c;
+}
+
+static uint16_t compress_counter(uint16_t c) {
+  if (((c ^ (c << 1)) & 0xAAAA) != 0xAAAA) {
+    handle_fault("ill-formed counter");
+  }
+  c = c & 0x5555;
+  c = ((c >> 1) | c) & 0x3333;
+  c = ((c >> 2) | c) & 0x0f0f;
+  c = ((c >> 4) | c) & 0x00ff;
+  return c;
+}
+
+static secbool pin_get_fails(uint32_t *ctr) {
+  const void *logs = NULL;
+  uint16_t len = 0;
+
+  wait_random();
+
+  if (sectrue != norcow_get(PIN_LOGS_KEY, &logs, &len) ||
+      len != PIN_LOG_HALFWORDS * sizeof(uint16_t)) {
+    handle_fault("no PIN logs");
+    return secfalse;
+  }
+
+  uint16_t c = compress_counter(((uint16_t *)logs)[0]);
+
+  secbool success = secfalse;
+
+  uint16_t correct_bytes_cnt = 0;
+
+  for (uint8_t i = 0; i < PIN_LOG_HALFWORDS; i++) {
+    wait_random();
+    correct_bytes_cnt += compress_counter(((uint16_t *)logs)[i]) == c;
+    *ctr = c;
+  }
+
+  success = sectrue * (correct_bytes_cnt == PIN_LOG_HALFWORDS);
+
+  return success;
```

Comment:
I think the variable `success` is not need anymore:
```c
  return sectrue * (correct_bytes_cnt == PIN_LOG_HALFWORDS);
```

Reply:
I'd say
```c
  if (correct_bytes_cnt != PIN_LOG_HALFWORDS) {
    handle_fault(...);
    return secfalse;
  }
  return sectrue;
```

Reply:
I agree that what Andrew suggests is more intuitive, but I think the idea behind the current solution was to avoid conditional branching if it's not too complicated to do so.

Reply:
yes that was the intention. 

removed the variable: https://github.com/trezor/trezor-firmware/pull/3228/commits/e3e6fb962eac98a4896a5649c25c4d26988517e5

Reply:
1. Unused variable left behind.
2. We should be calling `handle_fault()` if there is an inconsistency in the storage, which is the case when `correct_bytes_cnt != PIN_LOG_HALFWORDS`.

Reply:
https://github.com/trezor/trezor-firmware/commit/620aeb4271e7aac55156ea1bf30b886558cc57b7